code_samples/register_actions/src/org/jetbrains/tutorials/actions/SimplePopDialogAction.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.tutorials.actions;
+
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.ui.Messages;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.*;
+
+/**
+ * Action class to demonstrate how to interact with the IntelliJ Platform.
+ * The only action this class performs is to provide the user with a popup dialog as feedback.
+ * Typically this class is instantiated by the IntelliJ Platform framework based on declarations
+ * in the plugin.xml file. But when added at runtime this class is instantiated by an action group.
+ * @see com.intellij.openapi.actionSystem.AnAction
+ * @see com.intellij.openapi.actionSystem.AnActionEvent
+ */
+public class SimplePopDialogAction extends AnAction {
+
+  // Capture the generic message icon for display in the dialog.
+  private static final Icon ourIcon = Messages.getInformationIcon();
+
+  /**
+   * This default constructor is used by the IntelliJ Platform framework to
+   * instantiate this class based on plugin.xml declarations. Only needed in SimplePopDialogAction
+   * class because another constructor is overridden.
+   * @see AnAction#AnAction()
+   */
+  public SimplePopDialogAction() {
+    super();
+  }
+
+  /**
+   * This constructor is used to support dynamically added menu actions.
+   * It sets the text, description to be displayed for the menu item.
+   * Otherwise, the default AnAction constructor is used by the IntelliJ Platform.
+   * @see AnAction#AnAction(String, String, Icon)
+   * @param menuText  The text to be displayed as a menu item.
+   * @param menuDescription  The description of the menu item.
+   * @param menuIcon  The icon to be used with the menu item.
+   */
+  public SimplePopDialogAction(@Nullable String menuText, @Nullable String menuDescription, @Nullable Icon menuIcon) {
+    super(menuText, menuDescription, menuIcon);
+  }
+
+  /**
+   * Gives the user feedback when the dynamic action menu is chosen.
+   * Pops a simple message dialog.
+   * @param anActionEvent Event received when the associated menu item is chosen.
+   */
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
+    // Using the event, create and show a dialog
```

Comment:
I actually like Anna's original example more in this place. It demonstrated accessing data from `AnActionEvent`, which is a fundamental part of using actions, and your example shows only working with the action presentation, and does that in a way which is unlikely to be useful for plugin developers.

Reply:
I like Anna's `anAction` implementation as well. Here's my thinking...
IMO a tutorial like `register_actions` should focus on the ways actions/groups get registered and how they impose context requirements. (`psi_demo` already does a nice job of demonstrating data access from `AnActionEvent`.)
`SimplePopDialogAction` demonstrates determining context, and with `CustomDefaultActionGroup` shows how to layer more restrictive contexts. Maybe there are more relevant contexts; I'm open to suggestions. That said, there's virtue in `SimplePopDialogAction` having modest context requirements because it can be used/reused throughout the code sample. Previously most of the example actions in `register_actions` did not give  users feedback when a menu was chosen. I think providing user feedback (however trivial) is important in this tutorial, and `register_actions` now uses `SimplePopDialogAction` to do that in all cases.

Reply:
Providing user feedback is definitely better than having empty actions as in Anna's version of the tutorial. However, I do think that working with action data is an essential part of working with actions, and should be demonstrated in the core actions tutorial, not in a more topic-specific tutorial such as `psi_demo`. You can easily combine the two and show some data taken from the actionevent in the dialog.

Reply:
I have committed what I hope is a good compromise - using the `Navigatable` object opportunistically in `SimplePopDialogAction.actionPerformed()`. I have updated the  _Registering an Action_ documentation to discourage using class fields in `AnAction` subclasses.